### PR TITLE
Add support for minting skill, contract and connection packages

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR adds support for minting skill, contract and connection packages and test for the same

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: autonomy/cli/mint.py [hunk 1]
```diff
@@ -94,3 +94,87 @@ def protocol(
...
+    ctx: Context,
+    package_path: Path,
+    keys: Path,
+    password: Optional[str],
+    dependencies: Tuple[str],
+    nft: Optional[str],
+) -> None:
+    """Mint a skill component."""
+
+    mint_component(
```
Comment (solarw) [inline autonomy/cli/mint.py:172]: minor optional comment

probably it's possible to make a command argument type of PackageConnection,cause subcommands diff are only in this argument. just DRY

**Review by solarw — APPROVED** [review]
LGTM

**Comment by angrybayblade** [comment]
@Adamantios @0xArdi When try to mint an agent package I get this

```
[2023-01-10 11:33:42,272][WARNING] Unable to estimate gas with default state , ContractLogicError: execution reverted: Error: VM Exception while processing transaction: reverted with custom error 'ZeroValue()'
[2023-01-10 11:33:42,279][WARNING] Unable to retrieve gas estimate: Could not encode to JSON: dict had unencodable value at keys: {'params': because (list had unencodable value at index: [1: because (Object of type _LiteralGenericAlias is not JSON serializable)])}
[2023-01-10 11:33:42,290][WARNING] Unable to send transaction: {'code': -32000, 'message': 'Transaction requires at least 22352 gas but got 1', 'data': {'message': 'Transaction requires at least 22352 gas but got 1'}}
Component minted with:
        Public ID: valory/hello_world:0.1.0
        Metadata Hash: 0xbdf28f382d1611f71845722f16662432a29ad8ae7268175330f7f21a5feac25c
Error: Could not verify metadata hash to retrieve the token ID
```

But the I can mint other components without having any issues. Any idea what might be causing this

File: autonomy/chain/exceptions.py [hunk 2]
```diff
@@ -29,4 +29,4 @@ class ComponentMintFailed(ChainInteractionError):
 
 
 class FailedToRetrieveTokenId(ChainInteractionError):
-    """Raise token ID retrieving fails for minted component."""
+    """Raise when token ID retrieving fails for minted component."""
```
Comment (angrybayblade) [inline autonomy/chain/exceptions.py:32]: Typo

File: autonomy/cli/helpers/chain.py [hunk 3]
```diff
@@ -41,12 +41,16 @@ def mint_component(  # pylint: disable=too-many-arguments
...
-    dependencies: Optional[List[int]] = None,
+    dependencies: List[int],
     nft_image_hash: Optional[str] = None,
     password: Optional[str] = None,
 ) -> None:
     """Mint component."""
 
+    is_agent = package_type == PackageType.AGENT
+    if is_agent and len(dependencies) == 0:
+        raise click.ClickException("Agent packages needs to have dependencies")
```
Comment (angrybayblade) [inline autonomy/cli/helpers/chain.py:52]: Agent packages needs to have at least one dependency

File: packages/valory/contracts/component_registry/contract.py [hunk 4]
```diff
@@ -29,7 +28,7 @@
 from aea.crypto.base import LedgerApi
 
 
-PUBLIC_ID = PublicId.from_str("valory/componentRegistry:0.1.0")
+PUBLIC_ID = PublicId.from_str("valory/component_registry:0.1.0")
```
Comment (angrybayblade) [inline packages/valory/contracts/component_registry/contract.py:31]: This should be in snake case

**Review by solarw — APPROVED** [review]
LGTM

File: tests/test_autonomy/test_cli/test_mint/test_mint_components.py [hunk 5]
```diff
@@ -46,24 +47,43 @@ def get_create_transaction(self, *args: Any, **kwargs: Any) -> None:
...
-            commands=("packages/valory/protocols/acn", str(ETHEREUM_KEY_DEPLOYER)),
-        )
+        if package_type == PackageType.AGENT:
+            result = self.run_cli(
+                commands=(
+                    package_type.value,
+                    package_path,
+                    str(ETHEREUM_KEY_DEPLOYER),
+                    "-d",
+                    "1",
```
Comment (DavidMinarsch) [inline tests/test_autonomy/test_cli/test_mint/test_mint_components.py:80]: no magic values

File: autonomy/chain/mint.py [hunk 6]
```diff
@@ -179,23 +166,49 @@ def mint_component(
...
+            metadata_hash=metadata_hash,
+            unit_type=component_type,
+        )
     except RequestsConnectionError as e:
         raise FailedToRetrieveTokenId(
             "Connection interrupted while waiting for the unitId emit event"
         ) from e
 
+
+def verify_and_fetch_token_id_from_event(
```
[the remark this answers is not in the document]
Comment (angrybayblade) [reply autonomy/chain/mint.py:192]: Addressed [ee07f24](https://github.com/valory-xyz/open-autonomy/pull/1719/commits/ee07f24a9b2ca44869eb0be162a474236905c7bf)

File: tests/test_autonomy/test_cli/test_helpers/test_chain_helpers.py [hunk 7]
```diff
@@ -61,6 +61,7 @@ def test_mint_component_rpc_connect_fail(
                 package_type=PackageType.PROTOCOL,
                 keys=ETHEREUM_KEY_DEPLOYER,
                 chain_type=ChainType.LOCAL,
+                dependencies=[],
```
[the remark this answers is not in the document]
Comment (angrybayblade) [reply tests/test_autonomy/test_cli/test_helpers/test_chain_helpers.py:64]: It's hard to design proper tests in the current situation, so I suggest we add proper tests once we're done with the entire command implementation. Till then we have an integration test to check if minting works for the components or not and some unit tests for the helper methods

File: tests/test_autonomy/test_cli/test_mint/test_mint_components.py (hunk 5 above)
Comment (angrybayblade) [reply tests/test_autonomy/test_cli/test_mint/test_mint_components.py:80]: This is a roughly put together test, I'll update the test properly once we're done with the command implementation

File: autonomy/chain/mint.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (angrybayblade) [reply autonomy/chain/mint.py]: Addressed

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**